### PR TITLE
Feature/merge ethereum

### PR DESCRIPTION
**Pull request by jiangshangqifeng**
merge ethereum

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -1173,51 +1194,151 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
...
 
-	// Iterate over the blocks and insert when the verifier permits
-	for i, block := range chain {
+	// Peek the error for the first block to decide the directing import logic
+	it := newInsertIterator(chain, results, bc.Validator())
+
+	block, err := it.next()
+	switch {
+	// First block is pruned, insert as sidechain and reorg only if TD grows enough
+	case err == consensus.ErrPrunedAncestor:
```
Comment (cheng762) [inline core/blockchain.go:1203]: this case seems no use,i suggest delete it

**Review by cheng762 — APPROVED** [review]
seems good

**Comment by niuxiaojie81** [comment]
SGTM

**Review by niuxiaojie81 — APPROVED** [review]

File: core/blockchain.go (hunk 1 above)
Comment (jiangshangqifeng) [reply core/blockchain.go]: done

**Pull request merged**